test/cli/utils_test.go

```diff
@@ -271,7 +274,137 @@ func getSyftBinaryLocation(t testing.TB) string {
 		// SYFT_BINARY_LOCATION is the absolute path to the snapshot binary
 		return os.Getenv("SYFT_BINARY_LOCATION")
 	}
-	return getSyftBinaryLocationByOS(t, runtime.GOOS)
+	bin := getSyftBinaryLocationByOS(t, runtime.GOOS)
+
+	// only run on valid bin target, when not running in CI
+	if bin != "" && os.Getenv("CI") == "" {
+		buildSyft(t, bin)
+		// regardless if we have a successful build, don't attempt to keep building
+		_ = os.Setenv("SYFT_BINARY_LOCATION", bin)
+	}
+
+	return bin
+}
+
+func buildSyft(t testing.TB, outfile string) {
+	dir := repoRoot(t)
+
+	start := time.Now()
+
+	var stdout, stderr string
+	var err error
+	switch "go" {
+	case "go":
+		stdout, stderr, err = buildSyftWithGo(dir, outfile)
+	case "goreleaser":
+		stdout, stderr, err = buildSyftWithGoreleaser(dir)
+	case "make":
+		stdout, stderr, err = buildSyftWithMake(dir)
+	}
+
+	took := time.Now().Sub(start).Round(time.Millisecond)
+	if err == nil {
+		if len(stderr) == 0 {
+			t.Logf("binary is up to date: %s in %v", outfile, took)
```

Comment:
Note this is not entirely accurate. This is reported if there are no code changes and the build cache hasn't changed, but the binary is rebuilt anyway.